gc/base/HeapLinkedFreeHeader.hpp

```diff
@@ -178,47 +157,25 @@ class MM_HeapLinkedFreeHeader
 	 */
 	MMINLINE MM_HeapLinkedFreeHeader* afterEnd() { return (MM_HeapLinkedFreeHeader*)( ((uintptr_t)this) + getSize() ); }
 
-	/**
-	 * Mark the specified region of memory as all single slot holes
-	 * @param[in] addrBase the address to start marking from
-	 * @param freeEntrySize the number of bytes to be consumed
-	 */
-	MMINLINE static void
-	fillWithSingleSlotHoles(void* addrBase, uintptr_t freeEntrySize)
-	{
-#if defined(SPLIT_NEXT_POINTER)
-		uint32_t *freeSlot = (uint32_t *) addrBase;
-		while(freeEntrySize) {
-			*freeSlot++ = J9_GC_SINGLE_SLOT_HOLE;
-			freeEntrySize -= sizeof(uint32_t);
-		}
-#else /* defined(SPLIT_NEXT_POINTER) */
-		uintptr_t *freeSlot = (uintptr_t*) addrBase;
-		while(freeEntrySize) {
-			*freeSlot++ = J9_GC_SINGLE_SLOT_HOLE;
-			freeEntrySize -= sizeof(uintptr_t);
-		}
-#endif /* defined(SPLIT_NEXT_POINTER) */
-	}
-
 	/**
 	 * Mark the specified region of memory as walkable dark matter
 	 * @param[in] addrBase the address where the hole begins
 	 * @param[in] freeEntrySize the number of bytes to be consumed (must be a multiple of sizeof(uintptr_t))
 	 * @return The header written or null if the space was too small and was filled with single-slot holes
 	 */
 	MMINLINE static MM_HeapLinkedFreeHeader*
-	fillWithHoles(void* addrBase, uintptr_t freeEntrySize)
+	fillWithHoles(void* addrBase, uintptr_t freeEntrySize, bool compressed)
 	{
-		MM_HeapLinkedFreeHeader *freeEntry = NULL;
+		MM_HeapLinkedFreeHeader *freeEntry = (MM_HeapLinkedFreeHeader *)addrBase;
 		if (freeEntrySize < sizeof(MM_HeapLinkedFreeHeader)) {
-			/* Entry will be abandoned. Recycle the remainder as single slot entries */
-			fillWithSingleSlotHoles(addrBase, freeEntrySize);
+			if (0 != freeEntrySize) {
+				/* Assert_MM_true(freeEntrySize == sizeof(uintptr_t)); */
+				freeEntry->setNextImpl(J9_GC_SINGLE_SLOT_HOLE, compressed);
```

Comment:
Assuming compressed references,

* size of a single slot hole is 8,
* size of a full hole is 16,
* size of the class-ptr/next-ptr is 4.

It looks like we would write the `J9_GC_SINGLE_SLOT_HOLE` flag into each half of the single-slot-hole (which is, every 4 bytes of the free memory range). With this PR, we're only writing to the first half. I don't understand why we would write to the second half, so I don't know if it's safe to change.

Reply:
It caused no failures in a full sanity build, and I checked the code - single slot hole is assumed to be uintptr_t size.

Reply:
Good enough for me 👍